#' ---
title: 'fcoex: co-expression for single-cell data'
author:
- name: Tiago Lubiana
  affiliation: Computational Systems Biology Laboratory, University of SÃ£o Paulo, Brazil
output:
  html_document:
    toc: yes
  pdf_document:
    toc: yes
  prettydoc::html_pretty:
    highlight: github
    theme: cayman
package: fcoex
vignette: > 
  %\VignetteIndexEntry{fcoex: co-expression for single-cell data}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

The goal of fcoex is to provide a simple and intuitive way to generate co-expression nets and modules for single cell data. It is based in 3 steps:

- Pre-processing and label assignement (prior to fcoex)
- Discretization of gene expression
- Correlation and module detection via the FCBF algorithm (Fast Correlation-Based Filter)

First of all, we will a already preprocessed single cell dataset from 10XGenomics ( preprocessed according to https://osca.bioconductor.org/a-basic-analysis.html#preprocessing-import-to-r, 14/08/2019). It contains peripheral blood mononuclear cells and the most variable genes.

```{r  Loading datasets, message=FALSE  }
library(fcoex, quietly = TRUE)
library(SingleCellExperiment, quietly = TRUE)
data("mini_pbmc3k")
mini_pbmc3k
```

Now let's use the normalized data and the cluster labels to build the co-expresison networks.
The labels were obtained by louvain clustering on a graph build from nearest neighbours. That means that these labels are a posteriori, and this depends on the choice of the analyst. 

The fcoex object is created from 2 different pieces: a previously normalized expression table (genes in rows) and a target factor with classes for the cells. 

```{r Creating fcoex object, message=FALSE }
target <- colData(mini_pbmc3k)$clusters
exprs <- as.data.frame(assay(mini_pbmc3k, 'logcounts'))

fc <- new_fcoex(data.frame(exprs),target)

```

The first step is the conversion of the count matrix into a discretized dataframe. The standar of fcoex is a simple binarization that works as follows:

For each gene, the maximum and minimum values are stored. 
This range is divided in n bins of equal width (parameter to be set).
The first bin is assigned to the class "low" and all the others to the class "high".

```{r Discretizing dataset, message=FALSE }
fc <- discretize(fc, number_of_bins = 8)
```

Note that many other discretizations are avaible, from the implementations in the FCBF Bioconductor package. This step affects the final results in many ways. However, we found empirically that the default parameter often yields interesting results. 

After the discretization, we proceed to constructing a network and extracting modules. The co-expression adjacency matriz generated is modular in its inception. All correlations used are calculated via Symmetrical Uncertainty. Three steps are present:

1 - A filter of n genes to be considered, ranked by correlation to the target variable. 

2 - Detection of predominantly correlated genes, a feature selection approach defined in the FCBF algorithm

3 - Building of modules around selected genes. Correlations between two genes are kept if they are more correlated to each other than to the target lables

You can choose either to have a non-parallel processing, with a progress bar, or a faster parallel processing without progress bar. Up to you. 

```{r Finding cbf modules, message=FALSE }
fc <- find_cbf_modules(fc,n_genes = 200, verbose = FALSE, is_parallel = FALSE)
```


There are two functions that do the same: both get_nets and plot_interactions take the modules and plot networks. You can pick the name you like better. These visualizations were heavily inspired by the CEMiTool package, as much of the code in fcoex.

We will take a look at the first two networks 

```{r Plotting module networks, message=FALSE }
fc <- get_nets(fc)

# Example plot 
print(show_net(fc)[1])
print(show_net(fc)[2])
```

To save the plots, you can run the save plots function, which will create a "./Plots" directory and store plots there.
```{r Saving plots, message=FALSE, results='hide'}
save_plots(name = "fcoex_vignette", fc,force = TRUE)
```

You can also run  over-representation analysis to see if the modules correspond to any known biological pathway. For this, we will use the reactome groups available in the CEMiTool package:

```{r Running ORA analysis, message=FALSE}
gmt_fname <- system.file("extdata", "pathways.gmt", package = "CEMiTool")
gmt_in <- read_gmt(gmt_fname)
fc <- mod_ora(fc, gmt_in)
fc <- plot_ora(fc)
print(show_ora(fc)[1])
print(show_ora(fc)[2])
```

Now we can save the plots again. Note that we have to set the force parameter  equal to TRUE now, as the "./Plots" directory was already created in the previous step. 

```{r Saving plots again, message=FALSE, results='hide'}
save_plots(name = "fcoex_vignette", fc, force = TRUE)
```



There is now a folder with the correlations, to explore the data. 

We will use the module assignments to subdivide the cells in populations of interest. This is a way to explore the data and look for possible novel groupings ignored in the previous clustering step.

```{r Reclustering, , message=FALSE}

fc <- recluster(fc)

```

We generated new labels based on each fcoex module. Now we will visualize them using UMAP
```{r Visualizing}

colData(mini_pbmc3k) <- cbind(colData(mini_pbmc3k), mod_FCER1G = idents(fc)$FCER1G)
colData(mini_pbmc3k) <- cbind(colData(mini_pbmc3k), mod_CD3D = idents(fc)$CD3D)

library(scater)
# Let's see the original clusters
plotReducedDim(mini_pbmc3k, dimred="UMAP", colour_by="clusters")

library(gridExtra)
# And let's see the population represented in the modules CD3D and FCER1G.
# Notably, the patterns are largely influenced by the patterns of header genes
p1 <- plotReducedDim(mini_pbmc3k, dimred="UMAP", colour_by="mod_FCER1G")
p2 <- plotReducedDim(mini_pbmc3k, dimred="UMAP", colour_by="FCER1G")
p3 <- plotReducedDim(mini_pbmc3k, dimred="UMAP", colour_by="mod_CD3D")
p4 <- plotReducedDim(mini_pbmc3k, dimred="UMAP", colour_by="CD3D")

grid.arrange(p1, p2, p3, p4, nrow=2)
```

