---
title: 'fcoex: Co-expression for single-cell data'
output:
  html_document:
    toc: yes
  pdf_document:
    toc: yes
  prettydoc::html_pretty:
    highlight: github
    theme: cayman
    toc: yes
  vignette: > 
  %\VignetteIndexEntry{CEMiTool: Co-expression Modules Identification Tool}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

The goal of fcoex is to provide a simple and intuitive way to generate co-expression nets and modules for single cell data. It is based in 3 steps:

- Pre-processing and label assignement (prior to fcoex)
- Discretization of gene expression
- Correlation and module detection via the FCBF algorithm (Fast Correlation-Based Filter)

First of all, we will load and preprocess a single cell dataset from 10XGenomics (adapted from https://osca.bioconductor.org/a-basic-analysis.html#preprocessing-import-to-r, 14/08/2019):

```{r}

library(SingleCellExperiment)
#BiocManager::install(c('TENxPBMCData', 'scater', 'scran'))
library(TENxPBMCData)

sce <- TENxPBMCData('pbmc3k')

## reassign rownames
rownames(sce) <- rowData(sce)[, "Symbol_TENx"]

# subset for brevity
set.seed(3)
sce <- sce[hvg_genes,sample(ncol(sce),600)]
## counts dupes from top to bottom to make a logical vector and remove-it
dupes <- duplicated(rownames(sce))

sce <- sce[!dupes, ]

# normalize data
sce <- scater::normalize(sce)

#get variable genes and run PCA
library(scran)
fit <- trendVar(sce, use.spikes = FALSE)
dec <- decomposeVar(sce, fit)
dec <- dec[order(dec$bio, decreasing = TRUE), ] # order by bio var
hvg_genes <- rownames(dec)[dec$bio > 0]
metadata(sce)$hvg_genes <- hvg_genes
sce <- scater::runPCA(sce, ncomponents = 50,
              feature_set = hvg_genes)

# get clusters
set.seed(1234) # to make results reproducible
snng <- buildSNNGraph(sce, k = 5, use.dimred = "PCA")
snng_clusters <- igraph::cluster_louvain(snng)
colData(sce)$clusters <- as.factor(snng_clusters$membership)


#visualize the data
library(scater)
sce <- runUMAP(sce, use_dimred="PCA")
plotReducedDim(sce, use_dimred="UMAP", colour_by="clusters")
```

Now let's use the normalized data and the cluster labels to build the co-expresison networks!

```{r}
target <- colData(sce)$clusters
exprs <- as.data.frame(assay(sce, 'logcounts'))

library(fcoex)
fc <- new_fcoex(data.frame(exprs),target)
fc <- discretize(fc, number_of_bins = 8)
fc <- find_cbf_modules(fc,n_genes = 200, verbose = FALSE, is_parallel = TRUE)
fc <- plot_interactions(fc)
gmt_fname <- system.file("extdata", "pathways.gmt", package = "CEMiTool")
gmt_in <- read_gmt(gmt_fname)

fc <- mod_ora(fc, gmt_in)
fc <- plot_ora(fc)
save_plots(name = "fcoex_vignette", fc, force = TRUE)
```

There is now a folder with the correlations, to explore the data. 

We will use the module assignments to subdivide the cells in populations of interest. This is a way to explore the data and look for possible novel groupings ignored in the previous clustering step.



```{r}

fc <- recluster(fc)

```

We generated new labels based on each fcoex module. Now we will visualize them using UMAP
```{r}

sce@colData <- cbind(colData(sce), mod_FCER1G = fc@mod_idents$FCER1G)
sce@colData <- cbind(colData(sce), mod_CD3D = fc@mod_idents$CD3D)
# Let's see the original clusters
plotReducedDim(sce, use_dimred="UMAP", colour_by="clusters")

# And let's see the population represented in the modules CD3D and FCER1G.
# Notably, the patterns are largely influenced by the patterns of header genes
plotReducedDim(sce, use_dimred="UMAP", colour_by="mod_FCER1G")
plotReducedDim(sce, use_dimred="UMAP", colour_by="FCER1G")
plotReducedDim(sce, use_dimred="UMAP", colour_by="mod_CD3D")
plotReducedDim(sce, use_dimred="UMAP", colour_by="CD3D")

mini_pbmc3k <- sce
save( file =  "data/mini_pbmc3k.rda", list =  "mini_pbmc3k")

```


AS
